src/Sarif.Multitool/Rules/SARIF2001.AuthorHighQualityMessages.cs

```diff
@@ -3,85 +3,106 @@
+            nameof(RuleResources.SARIF2001_AuthorHighQualityMessages_Warning_TerminateWithPeriod_Text)
         };
 
         public override FailureLevel DefaultLevel => FailureLevel.Warning;
 
-        public override string Id => RuleId.AuthorHighQualityMessages;
+        private static readonly Regex s_dynamicContentRegex = new Regex(@"{[0-9]+}", RegexOptions.Compiled | RegexOptions.CultureInvariant);
+        private static readonly Regex s_enquoteDynamicContentRegex = new Regex(@"'{[0-9]+}'", RegexOptions.Compiled | RegexOptions.CultureInvariant);
 
-        protected override IEnumerable<string> MessageResourceNames
+        protected override void Analyze(Tool tool, string toolPointer)
         {
-            get
+            if (tool.Driver != null)
             {
-                return new string[]
-                {
-                    nameof(RuleResources.SARIF2001_AuthorHighQualityMessages_Warning_TerminateWithPeriod_Text)
-                };
+                AnalyzeToolDriver(tool.Driver, toolPointer.AtProperty(SarifPropertyName.Driver));
             }
         }
 
-        protected override void Analyze(ReportingDescriptor reportingDescriptor, string reportingDescriptorPointer)
+        private void AnalyzeToolDriver(ToolComponent toolComponent, string toolDriverPointer)
         {
-            AnalyzeMessageStrings(reportingDescriptor.MessageStrings, reportingDescriptorPointer);
+            if (toolComponent.Rules != null && toolComponent.Rules.Count > 0)
+            {
+                foreach (ReportingDescriptor rule in toolComponent.Rules)
+                {
+                    AnalyzeReportingDescriptor(rule, toolDriverPointer.AtProperty(SarifPropertyName.Rules));
+                }
+            }
         }
 
-        private void AnalyzeMessageStrings(
-            IDictionary<string, MultiformatMessageString> messageStrings,
-            string reportingDescriptorPointer)
+        private void AnalyzeReportingDescriptor(ReportingDescriptor rule, string reportingDescriptorPointer)
         {
-            if (messageStrings != null)
+            if (rule.MessageStrings != null && rule.MessageStrings.Count > 0)
             {
                 string messageStringsPointer = reportingDescriptorPointer.AtProperty(SarifPropertyName.MessageStrings);
-
-                foreach (string key in messageStrings.Keys)
+                foreach (KeyValuePair<string, MultiformatMessageString> message in rule.MessageStrings)
                 {
-                    MultiformatMessageString messageString = messageStrings[key];
-
-                    string messageStringPointer = messageStringsPointer.AtProperty(key);
-
-                    AnalyzeMessageString(messageString.Text, messageStringPointer, SarifPropertyName.Text);
+                    AnalyzeMessageString(message.Value.Text, messageStringsPointer.AtProperty(message.Key));
```

Comment:
>message.Value.Text [](start = 41, length = 18)

You'll also need to pass the message.Key to `AnalyzeMessageString` because it will surely be part of the message string. For example, "The message string 'x' in rule 'y' has a problem..."